### PR TITLE
Lesson12

### PR DESCRIPTION
**Pull request by DesoPy**
Wohe work lesson 12

### CODE REVIEW

File: Lesson12.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+If all nums inside the list are positive, execute the first function on that list and return the result of it.
+Otherwise, return the result of the second one
+"""
+
+
+def choose_func(nums: list, func1, func2):
+    if reduce(lambda x, y: abs(x + y), nums) == reduce(lambda x, y: x + y, nums):
+        return_func = func1(nums)
+        return return_func
+    elif reduce(lambda x, y: abs(x + y), nums) > reduce(lambda x, y: x + y, nums):
```
Comment (fadich) [inline Lesson12.py]: Redundant elif statement:
```suggestion
    return func2(nums)
```

File: Lesson12.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+Task 3
+
+Write a function called `choose_func` which takes a list of nums and 2 callback functions. 
+If all nums inside the list are positive, execute the first function on that list and return the result of it.
+Otherwise, return the result of the second one
+"""
+
+
+def choose_func(nums: list, func1, func2):
+    if reduce(lambda x, y: abs(x + y), nums) == reduce(lambda x, y: x + y, nums):
```
Comment (fadich) [inline Lesson12.py]: To complex expression. I'd suggest something like:
`all(map(lambda x: x >= 0, nums))`
or
`len([i for i in nums if i < 0])`
or
`len(tuple(filter(lambda x: x < 0, a)))`
and so on...

Or just to use regular for-loop above. :)

File: guess_the_words/guess_the_word.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        choose_loop(choose_user, player_name)
+
+        question_for_user = input('Do you want continue [Y/N]?')
+
+        if question_for_user.lower() not in ['y', ' ']:
+            print('Program was closed')
+            break
+        else:
+            continue
+except (Exception, SystemExit, KeyboardInterrupt, GeneratorExit):
```
Comment (fadich) [inline guess_the_words/guess_the_word.py:20]:
```suggestion
except BaseException:
```

or just
```
except:
    ...
```

File: guess_the_words/function.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                break
+
+        else:
+            print('Incorrect input.')
+            break
+
+
+def val_alpha(player_letters: list) -> list:
+    if set(player_letters) - set(string.ascii_lowercase):
+        print("Unsupported symbols, use letters instead")
```
Comment (fadich) [inline guess_the_words/function.py:82]:
```suggestion
```

File: guess_the_words/function.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    count_dict = {'Date time game': '', 'Player name': player_name, 'Game result': 'Lose',
+                  'Score': {'Count tries': 0, 'Score points': 0}}
+
+    target_word = random.choice(create_list())
+
+    i = 0
+    while i < 6:
+        player_letters = list(input(f'Attempt {i + 1} of 6: ').lower())
+
+        if val_alpha(player_letters):
```
Comment (fadich) [inline guess_the_words/function.py:106]:
```suggestion
        if val_alpha(player_letters):
            print("Unsupported symbols, use letters instead")
```

File: guess_the_words/function.py (hunk 4 above)
Comment (fadich) [reply guess_the_words/function.py:82]: It's better to move it into `body_loop()`

**Review by fadich — APPROVED** [review]
Suggesting minor fixes, but in general it looks very nice. Well done!

File: Lesson12.py (hunk 1 above)
Comment (DesoPy) [reply Lesson12.py]: Corrected, you were just talking about it today, updated in the Lesson12 brunch

File: Lesson12.py (hunk 2 above)
Comment (DesoPy) [reply Lesson12.py]: I don’t even know why I made such a long record) I changed it to one of the proposed options - it works the same way)) updated in the Lesson12 brunch

File: guess_the_words/function.py (hunk 4 above)
Comment (DesoPy) [reply guess_the_words/function.py:82]: So i did, i removed that check and moved it to body_loop

File: guess_the_words/function.py (hunk 5 above)
Comment (DesoPy) [reply guess_the_words/function.py:106]: I changed logic for:

    while i < 6:
        player_letters = list(input(f'Attempt {i + 1} of 6: ').lower())

        if set(player_letters) - set(string.ascii_lowercase):
            print("Unsupported symbols, use letters instead")
            continue

        if len(player_letters) > 5:
            print('You enter word more then 5 letters')
            continue
        elif len(player_letters) < 5:
            print('You enter word less then 5 letters')
            continue

File: guess_the_words/function.py (hunk 4 above)
Comment (DesoPy) [reply guess_the_words/function.py:82]: Code was changed, correctly on link https://github.com/DesoPy/Beetroot_Home_Work/blob/Guess_the_word/guess_the_words/function.py

**Review by fadich — APPROVED** [review]

**Pull request merged**